File: 234-Palindrome-Linked-List/go/main.go
package main

import (
	"fmt"
)

func main() {
	fmt.Println("Hello World")
	first := ListNode{
		Val:  1,
		Next: nil,
	}

	second := ListNode{
		Val:  2,
		Next: &first,
	}
	third := ListNode{
		Val:  2,
		Next: &second,
	}
	head := ListNode{
		Val:  1,
		Next: &third,
	}

	fmt.Println(isPalindrome(&head))

}

type ListNode struct {
	Val  int
	Next *ListNode
}

func isPalindrome(head *ListNode) bool {
	var vecs []int
	for head != nil {
		fmt.Println(head.Val)
		vecs = append(vecs, head.Val)
		head = head.Next
	}

	first := 0
	length := len(vecs)
	last := length - 1

	for i := 0; i < length/2; i++ {
		if vecs[first] != vecs[last] {
			return false
		}
		first++
		last--
	}
	return true
}

File: 234-Palindrome-Linked-List/rs/src/main.rs

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
  pub val: i32,
  pub next: Option<Box<ListNode>>
}

impl ListNode {
  #[inline]
  fn new(val: i32) -> Self {
    ListNode {
      next: None,
      val
    }
  }
}

fn main() {
    let first = ListNode::new(1);
    
    
    let sec = ListNode{
        val: 2,
		next: Some(Box::new(first)),
    };
	
	let third = ListNode{
        val: 2,
		next: Some(Box::new(sec)),
    };
	let head = ListNode{
        val: 1,
		next: Some(Box::new(third)),
    };
	
	let answer = is_palindrome(Some(Box::new(head)));
	println!("{:?}", answer);
}




pub fn is_palindrome(mut head: Option<Box<ListNode>>) -> bool {

    let mut head_vec: Vec<i32> = vec![] ;
    
    while let Some(node) = head {
        head_vec.push(node.val);
        head = node.next;
    }
    
    let mut first = 0;
    let mut length = head_vec.len();
    // let length = head_vec.len();
    let mut last = length-1;
    
    let mut i = 0;
    let loop_end = length/2;
    while i < loop_end {
        if head_vec[first] != head_vec[last]{
            return false;
        }
        first += 1;
        last -= 1;
        i += 1;
    }
    true
}
